File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int search(TreeNode* root, int& K) {
if (root == NULL) {
return -1;
}
int leftVal = search(root->left, K);
if(leftVal != -1) {
return leftVal;
}
K--;
if(K == 0) {
return root->val;
}
return search(root->right, K);
}
int kthSmallest(TreeNode* root, int k) {
return search(root, k);
}

};

File: 230-kth-smallest-element-in-a-bst/kth-smallest-element-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode n;
    int k;
    
    
    public void getKthSmallestUsingInorder(TreeNode node) {
        if (node == null)
            return;
        
        
        getKthSmallestUsingInorder(node.left);
        
        
        this.k--;
        
        if (this.k == 0)
            this.n = node;
        
        
        if (this.k > 0)
            getKthSmallestUsingInorder(node.right);
    }
    
    
    public int kthSmallest(TreeNode root, int k) {
        this.k = k;
        getKthSmallestUsingInorder(root);
        return this.n.val;
    }
}